### ISSUE TITLE
Generic name for FALLBACK and CALL-ME

### ISSUE THREAD

**Issue by JJ** [Raku/problem-solving#119]
These two methods basically add syntactic sugar to objects and classes: in one case, they can create on the fly methods that do not exist, in the other give an object the power to use () so that it becomes a functor. As far as I know, there's no generic name for this kind of thing, but then, there was no generic name for phasers before they were introduced in Perl.
Giving stuff a name helps, among other things, to document it (which is why I gave it that label), but also creates a concept that can be later on filled with other things.
So as long as we have agreed that we *need* a name, what would that name be?
* Syntaxers (because they add syntax)
* Enhancers (because they enhance the object with new syntax)
* Shapers (they change the shape so that it can be used in new ways)
* Rakers (well, this is a long shot, but they "rake" as in a Japanese garden to give them new paths, and also they might be a bit peculiar to Raku, so...)

Any other idea?

> This comes from perl6/doc#3062, that asks for some documentation

**Comment by vrurg** [comment]
A common term for just two entities? Does it really makes sense? Besides, the only thing they have in common is that they somehow modify class behavior.

Though, if choosing among the option would be mandatory, I'd go for _shapers_ or _modifiers_.

**Comment by JJ** [comment]
El vie., 18 oct. 2019 a las 16:29, Vadim Belman (<notifications@github.com>)
escribió:

> A common term for just two entities? Does it really makes sense? Besides,
> the only thing they have in common is that they somehow modify class
> behavior.
>

Well, besides being methods that you can add to any class, just like
phasers. Actually, the term should include also phasers, or even things
like AT-KEY which don't have a name yet.

> Though, if choosing among the option would be mandatory, I'd go for
> *shapers* or *modifiers*.
>
modifiers should go with some qualifiers. So, anyone else for shapers?

**Comment by tmtvl** [comment]
Well, FALLBACK is practically a type of exception handling for invalid method call exceptions. It doesn't really change the shape of the object.
CALL-ME is part of the Callable role, it's what makes the object that is a subroutine behave the way it does.
So I don't really see the relation between them, apart from both being methods that get called discreetly rather than overtly.

**Comment by vendethiel** [comment]
Python and PHP call them "magic methods". (Though Python also calls them "Dunder" methods because they have a special naming convention).

**Comment by JJ** [comment]
El vie., 18 oct. 2019 a las 18:37, Tim Van den Langenbergh (<
notifications@github.com>) escribió:

> Well, FALLBACK is practically a type of exception handling for invalid
> method call exceptions. It doesn't really change the shape of the object.
> CALL-ME is part of the Callable role, it's what makes the object that is a
> subroutine behave the way it does.
>

It's that, but you can create CALL-ME on any object to give it the callable
syntax. That's why one of the proposed names was "syntaxer"

**Comment by vrurg** [comment]
`AT-KEY`, `AT-POS` make a good company to the rest.

Why _syntaxers_ isn't a good one is because `FALLBACK` doesn't actually change the syntax of using a class. It rather responsible for class behavior.

Why I like _shapers_ is because those methods rather give the final "shape" to a class. Yet, I still have a feeling that there is better term for them.

@JJ, if the right term is chosen are there plans for a corresponding section in the docs?

**Comment by JJ** [comment]
@vrurg it would go to the glossary and then of course it would have a category in the documentation index and maybe a specific section.

**Comment by coke** [comment]
Looking back through the comments on this ticket, I see a question asked about why we need a common name for these two items, but not a response that answers that particular question.

Without a driving force to create this, I would recommend we close this ticket.

**Comment by raiph** [comment]
Hi @coke,

I'd like to "take" this ticket as the "driver" and see how I do.

If there's some particular thing I'm supposed to do to "take" this ticket other than just progress it as I see fit, please let me know what that is in as brief a way as you can manage. TIA.

Feel free to skip reading the rest of this comment, and indeed ignore this issue from here on, at least until I produce an associated PR or @mention you.

----

My proposal is that I'll either close this issue with no action (ie act as you recommend if there's no "driving force") or close it with:

* A new term and entry added to the glossary;

* New uses of the term in at least three other places, namely pages documenting FALLBACK, CALL-ME and phasers.

I could do more, such as using the term on more pages, adding it as a category, and so on, but currently think I'll postpone such things until I've closed this issue.

> I see a question asked about why we need a common name for these two items

To be fair, JJ wrote:

> Actually, the term should include also phasers, or even things like AT-KEY which don't have a name yet.

Again, feel free to skip reading the rest of this comment. It's just my discussion of the terms "magic methods" and "hooks" and a concluding strawman proposal that I plan to act on by default if no one pipes up in the next week or so.

# "Magic methods"

Putting aside phasers, I'd say Vendethiel was spot on, in that Python and PHP use the term "magic methods" for exactly the kinds of methods JJ mentioned (before JJ added phasers to the mix).

So perhaps we adopt the same term "magic methods" (and skip lumping phasers in with magic methods, because that would be confusing).

Then again, googles like ["avoid magic methods"](https://www.google.com/search?q=%22avoid+magic+methods%22) suggest that enough devs have negative associations with "magic methods" that perhaps we should *avoid* using their term.

Aiui the negative association is due to problems with the *specifics* of particular magic methods in other PLs, not the general technique. But if Raku sticks with the same *term* I think folk will apply their negative associations to Raku even if it makes infinitely better use of the technique than Python. So I'm thinking the right solution here will be to introduce a new term, and mention magic methods in Python/PHP (because that *is* a good reference point) but find a way to do that without dragging in undue negative associations.

# Hooks

The traditional fully generic term is ["hooks"](https://en.wikipedia.org/wiki/Hooking).

This term has the upsides that it avoids the word "magic", and would naturally cover phasers too, and could cover other hooks in Raku too (eg traits).

----

Then again, "hooks" might be too broad, perhaps *far* too broad.

For starters, it refers to hooks in any software system, not just a programming language or PL related artifact such as a library or API.

But even if one ignores that, I think it would essentially *force* us to lump in any other Raku feature that is, after all, a hook. If it walks and talks like a hook, it will be hard to deny it's a hook. Definitely traits, for example. Any others you can think of?

----

I could extend the word search to come up with something narrower than hooks. But I'm currently thinking "hooks" might work well, and this comment is already long. So...

# What I [strawman propose](https://en.wikipedia.org/wiki/Straw_man_proposal)

* The word is "hooks".

* The glossary description is:

> Hook. Raku contains a broad range of [hooks](https://en.wikipedia.org/wiki/Hooking) that let you write custom handling that gets automatically invoked as Raku processes particular aspects of code. Raku's hooks include equivalents of the "magic methods" / "dunders" in PLs like Python/PHP (but avoiding their downsides) plus features such as [Phasers](https://docs.raku.org/language/phasers) and [traits](https://docs.raku.org/language/traits).

* (Perhaps we define a subset of "hooks" that's "HOOKS". HOOKS would be Raku's ALLCAPS methods/functions/statements, i.e. Raku's equivalents to "magic methods" in Python/PHP, plus phasers.)

**Comment by JJ** [comment]
What about `AT-KEY`? Does not look so much a hook as a Magic method.

**Comment by raiph** [comment]
Hi @JJ,

Thanks for feedback. :)

`AT-KEY` is absolutely a magic method, as are `BUILD` and cousins (well, that strays into method or submethod of course), and even methods like `^parameterize`.

Otoh, phasers aren't methods.

We could perhaps include phasers if we made the generic term something like "magic code". "code" would generalize to cover not only non-method hooks like `CATCH { ... }`, but also non-routine non-function thunkish scenarios like `now - INIT now`, compile-time declaration hooks (traits), and, well, the term could cover any and all hooks. (I note you seem to be thinking of things as being magic methods *vs* hooks, but instead it's that magic methods are one of many kinds of hook, which is the generic term for "magic code" and any other, well, hook.)

That said, I'm currently leery of *any* term with the word "magic" in it due to how many devs I see apparently successfully propagating the meme that "magic" is a bad thing, period, without any attention paid to what the term refers to.

----

The traditional fully generic industry standard term for "magic code" was, for many decades, ["hooks"](https://en.wikipedia.org/wiki/Hooking).

One possibility I was mulling was having a Hooks page with sub-sections that include these:

* `ALLCAPS` hooks. This covers hooks in Raku that are ALLCAPS methods/functions/statements/thunks, encompassing Raku's ALLCAPS "magic methods" as well as its uppercase phasers.)

* MOP `^hooks`. This covers MOP hooks such as `^parameterize`.

* Compile-time traits. This covers stuff like trait mods, including `will` which also involves phasers.

Or perhaps N separate pages that are all somehow tagged with a `hook` category?

**Comment by lizmat** [comment]
This issue will be discussed in the first Raku Resolutions meeting, see more information at https://dev.to/lizmat/raku-resolutions-17g7

**Comment by librasteve** [comment]
the approach I take for refactoring things (ie to recognise that there is a need for grouping) is

1. once is happenstance
2. twice is coincidence
3. three times is enemy action

**Comment by raiph** [comment]
**TL;DR** I propose we close this issue because the doc _already uses_ a term (in four places) that is (I claim) a perfect fit and we can just expand use of the existing term as and when we see fit.¹˒²

----

Here's what Wikipedia says the term (that the doc already uses) refers to:

> In computer programming, [this approach] is ... used to alter or augment the behaviour of [software] by intercepting [execution]. Code that handles such intercepted [execution] is called a ...

In the context of programming language features, these are not hacks by a programmer that the language is unaware of but rather _hooks_ that the language provides to allow principled use of user definable _handlers_.

While the proposed term can be understood to cover users writing arbitrary _user defined_ higher order functions that take functions as arguments that the function calls back, and also _builtin_ HOFs like `map` doing likewise, what we're really talking about here is use of the same concept _by the language itself_. Examples are Raku's `CATCH` blocks (a hook that a dev can hang exception handling code on) and `FALLBACK` methods (a hook that a dev can hang fallback handling code on).

And the term? Well, I already snuck it in above. If you missed it, you might want to read the Wikipedia page [Hooking](https://en.wikipedia.org/wiki/Hooking). 😁 BTW, one reason I think the term "Magic methods" is inappropriate is because many Raku hooks aren't methods.

----

Our doc already suitably uses the term "hook" in four places, ["In Raku, traits are compiler **hooks** ..."](https://docs.raku.org/language/traits), ["`for` actually **hooks** ..."](https://docs.raku.org/type/Iterable#:~:text=for%20actually%20hooks%20to%20the%20iterator%20role), ["... use `on-switch` **hook** ..."](https://docs.raku.org/language/perl-var#:~:text=you%20can%20use%20on%2Dswitch%20hook%20to%20reset%20line%20number%20on%20handle%20switch%2C%20and%20increment%20it%20manually), ["... a **hook** that executes whenever a value is retrieved from a container (`FETCH`) or when it is set (`STORE`)."](https://docs.raku.org/type/Proxy#:~:text=A%20Proxy%20is%20an%20object%20that%20allows%20you%20to%20set%20a%20hook%20that%20executes%20whenever%20a%20value%20is%20retrieved%20from%20a%20container%20(FETCH)%20or%20when%20it%20is%20set%20(STORE).).

These uses appear to assume a user will intuit what a "hook" is, partly, I think, because the term is indeed (imo) an intuitive generic name for what's going on.

----

I asked google's AI to pick one of three undefined terms given how devs typically react to them:

> Attitudes and understanding among devs in recent years associated with the programming terms "magic methods" vs "dunder methods" vs "hooks". Which of them would best be used in the vocabulary of a new programming language?

The full answer is [here](https://gist.github.com/raiph/d720ac21f1bcd4b84433ed5a73c54d39), but for convenience I'll excerpt here:

> **"Hooks"** is a widely understood and general term across many programming contexts ... It is consistently understood to mean a place in code that allows a developer to inject custom logic to augment or modify the behavior of a system at specific, predefined points ... Its broad and intuitive meaning makes it an excellent choice for new language vocabulary.
>
> ... the term "magic" can imply obscurity or non-explicit behavior ... This ambiguity makes it a less desirable formal term for a new language.
>
> ... **"Dunder methods"** ... describes *how* the methods are named, not their conceptual purpose as effectively as "hooks" ... A new language would likely want a term that describes the *functionality* clearly. 

----

Here's an off-the-top-of-my-head list of some Raku hooks:

* **Methods** `FALLBACK`, `CALL-ME`, `BUILD`, `TWEAK`, `FAILGOAL`, `FETCH`, `STORE`, `AT-KEY`, `AT-POS`, `.^parameterize`, `.WHAT`, `.HOW`, `.WHY`. (I think at least two of these last three are technically macro-like operators rather than methods but that's largely an implementation detail, and they're still hooks either way.) 

* **Non-methods** All the phasers and traits.

----

¹ Some readers may see my proposed framing and term as scope explosion. If so, please sleep on it, or, better, mull it until the next Raku Resolution meeting.

² If this proposal is more or less accepted I further propose we consider opening one related follow up doc issue, and perhaps more. The first and perhaps only issue for now would be to add the term "hook" to the doc's glossary. Other things to consider include: linking the four existing uses of the term to the glossary entry; starting a new page whose title is Hooks with a short stub explaining Raku's use of it; subsequent improvement as and when contributors feel inspired to work on it.

**Comment by lizmat** [comment]
In the Resolve-a-thon of 17 January 2026, it was felt that a generic term, especially "hook", would be inappropriate.

And also that we need to get a better classification of these features.

And since I've been working on an as yet unpublished project [Raku::Elements](https://github.com/lizmat/Raku-Elements?tab=readme-ov-file), it feels like a good idea to take on the task of writing a number of blog posts about the Raku features mentioned in this issue.

And to close this problem solving issue for now, in light of future new documentation issues.
